src/effcheck.ml

```diff
@@ -82,16 +86,32 @@ let rec tcheck env term =
         lst;
       (typ, eff)
     | _ -> Test.fail_report "tcheck: ListTrm must have a list type")
-  | Constructor (typ, name, payload_lst, eff) ->
-    (match check_opt_invariants (typ, name, payload_lst) with
-    | Ok _ ->
-      List.iter
-        (fun trm ->
-          (* ignore value but propatage error *)
-          tcheck env trm |> ignore)
-        payload_lst;
-      (typ, eff)
-    | Error msg -> Test.fail_report msg)
+  | Constructor (typ, descr, _, eff) ->
+    let tcheck_variant term =
+      match check_option_invars term with
+      | Ok _ -> (typ, eff)
+      | Error e -> failwith e
+    in
+    let tcheck_tuple = function
+      | Constructor (typ, TupleArity i, args, eff) ->
+        (* "tuple" term invariants:
+          1) arity i must be equal to length of args
+          2) typ must be [Tuple lst], where [lst] lists types that are same as types of args
+        *)
+        (* check invar (1) *)
+        if not (List.length args = i) then failwith "tcheck: arity invariant failed";
+        (* check invar (2) *)
+        (match typ with
+        | Tuple t_lst ->
+          if not (List.for_all2 (fun trm t -> types_compat (imm_type trm) t) args t_lst)
+          then failwith "tcheck: type mismatch"
+        | _ -> failwith "tcheck: cannot be reached");
+        (typ, eff)
+      | _ -> invalid_arg "Effcheck.tcheck.tcheck_tuple"
```

Comment:
You should use `Test.fail_report` for error-handling here, so that we get a properly formatted message from the main executable. (You can try making a mistake in this code and see how both approaches behave.)